src/main/java/com/faforever/client/api/TokenService.java

```diff
@@ -0,0 +1,104 @@
+package com.faforever.client.api;
+
+import com.faforever.client.preferences.PreferencesService;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.security.oauth2.common.OAuth2AccessToken;
+import org.springframework.stereotype.Service;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.Collections;
+
+@Service
+@Slf4j
+public class TokenService {
+  private final PreferencesService preferencesService;
+  private final RestTemplate restTemplate;
+  private OAuth2AccessToken tokenCache;
+
+  public TokenService(PreferencesService preferencesService) {
+    this.preferencesService = preferencesService;
+
+    restTemplate = new RestTemplateBuilder().
+        build();
+  }
+
+  @SneakyThrows
+  public OAuth2AccessToken getRefreshedToken() {
+    if (tokenCache == null || tokenCache.isExpired()) {
```

Comment:
Did you think about the case where the refreshtoken expired and you gotta log out the user 🤔

Reply:
Hmm would also be really rude tho logging him out during a game where being logged in is needed to communicate with the server. But I was once told Brutus plans on making it 30 days so frequent players will definitely experience that case. 🤔